### PR TITLE
Fix expand_tape_state_prep.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
[43708](https://app.shortcut.com/xanaduai/story/43708/lightning-ecosystem-devices-allow-stateprepbase-ops-to-be-used-mid-circuit) demands mid-circuit `StatePrep` support in Lightning. This PR fixes a small bug in `expand_tape_state_prep`.

**Description of the Change:**
`first_op`-bugfix

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by timmysilv — COMMENTED** [review]
nice find on that bug!

As for the template support - I'm worried that this helper function is getting too big for its own good with all these custom kwargs and overriden behaviours. What is the intention of this function? Should this decomposition apply to all qubit devices? I think we're only doing this because the old way of pre-processing was a little all over the place and because StatePrepBase decompositions haven't really been looked at in a while. I think this logic makes sense, but I don't think it belongs in _qubit_device.py

**Comment by vincentmr** [comment]
> nice find on that bug!
> 
> As for the template support - I'm worried that this helper function is getting too big for its own good with all these custom kwargs and overriden behaviours. What is the intention of this function? Should this decomposition apply to all qubit devices? I think we're only doing this because the old way of pre-processing was a little all over the place and because StatePrepBase decompositions haven't really been looked at in a while. I think this logic makes sense, but I don't think it belongs in _qubit_device.py

This PR comes about because of [SC-43708](https://app.shortcut.com/xanaduai/story/43708/lightning-ecosystem-devices-allow-stateprepbase-ops-to-be-used-mid-circuit). I wanted to add the following [test](https://github.com/PennyLaneAI/pennylane-lightning/pull/495/files#diff-c1389c82c8c9120903884cb7fc7d5bd2a81935a8272ec2d0b9ed59a94224cb48R1474) in Lightning, but I couldn't run the circuit or tape with `default.qubit`. I could not find a similar test in PL, so I thought I would add it and try to make it pass. If there is a better way to invoke and test mid-circuit state prep operations, would you please let me know. I agree this is hacky-washy as it is :D

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -272,10 +272,27 @@ def execute(self, circuit, **kwargs):
...
+                    qml.BasisState,
+                    qml.StatePrep,
+                    qml.BasisStatePreparation,
+                    qml.MottonenStatePreparation,
+                ),
+            )
+            for op in operations
+        ):
+            tape = qml.tape.QuantumTape(ops=operations, measurements=[])
+            tape = qml.tape.expand_tape_state_prep(tape, skip_first=True, force_decompose=True)
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/_qubit_device.py:289]: There may be a better spot, but I found this place to be a catch all. Without it, I get error like 
```
tests/tape/test_tape.py::TestExpand::test_expansion_state_prep[op0-decomp0-False-False] - pennylane._device.DeviceError: Gate BasisStatePreparation not supported on device default.qubit
```
This is also by analogy with Lightning where I put this filter in the `apply` function.

File: tests/tape/test_tape.py [hunk 2]
```diff
@@ -1021,35 +1022,57 @@ def test_depth_expansion(self):
...
 
         assert len(new_tape.operations) == len(true_decomposition)
         for tape_op, true_op in zip(new_tape.operations, true_decomposition):
             assert qml.equal(tape_op, true_op)
 
+        if skip_first:
+            dev = qml.device("default.qubit", wires=4)
+            new_tape = QuantumTape(new_tape)
+            ref_type = QuantumTape(true_decomposition)
+            assert np.allclose(dev.execute(new_tape), dev.execute(ref_type))
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply tests/tape/test_tape.py:1074]: > are they not the same if `skip_first` is False? 

No, the `true_decomposition` might have a leftover `BasisStatePreparation` which cannot be applied.

> also, is `new_tape` not a tape before this?

Yes, you're right.

**Review by timmysilv — APPROVED** [review]
thanks for separating out the bug!

**Comment by vincentmr** [comment]
> thanks for separating out the bug!

Turns out I was misusing `dev.execute` and mid-circuit state prep worked off the bat in Lightning too. All good now!

**Comment by Jaybsoni** [comment]
Clean 👍🏼

**Review by Jaybsoni — APPROVED** [review]
💯 Thanks for finding and fixing this 👍🏼

**Pull request merged**